File: src/main/java/dev/klomptech/jbark/parser/controlflow/InlineBodyParser.java
package dev.klomptech.jbark.parser.controlflow;

import dev.klomptech.jbark.errors.BarkError;
import dev.klomptech.jbark.lexer.Token;
import dev.klomptech.jbark.lexer.TokenType;
import dev.klomptech.jbark.parser.AstNode;
import dev.klomptech.jbark.parser.Keywords;
import dev.klomptech.jbark.parser.Parser;
import java.util.List;

// One-line then bodies: print, heel, again. No bury needed.
final class InlineBodyParser {

  private InlineBodyParser() {}

  static List<AstNode> parseThenSteps(
      final Parser parser, final int tokenCount, final int line, final String emptyError)
      throws BarkError {
    if (tokenCount <= 0 || parser.isAtEndOrLineEndOffset(0)) {
      throw new BarkError(line, emptyError);
    }
    Token first = parser.peekAt(0);
    if (first.is(TokenType.IDENTIFIER) && tokenCount == 1) {
      String word = parser.normalise(first.value());
      if (Keywords.BREAK_KEYWORDS.contains(word)) {
        parser.advanceBy(1);
        return List.of(new AstNode.Break(line));
      }
      if (Keywords.CONTINUE_KEYWORDS.contains(word)) {
        parser.advanceBy(1);
        return List.of(new AstNode.Continue(line));
      }
    }
    if (parser.lineHasWord(Keywords.PRINT_KEYWORDS)) {
      return List.of(parser.getPrintParser().parsePrintForNextTokens(tokenCount));
    }
    throw new BarkError(line, emptyError);
  }
}

File: pbark/pbark/parser/controlflow/inline_body_parser.py
from __future__ import annotations

from pbark.errors import BarkError
from pbark.lexer import TokenType
from pbark.parser import ast_node as AstNode
from pbark.parser import keywords as Keywords


class InlineBodyParser:
    @staticmethod
    def parse_then_steps(parser, token_count: int, line: int, empty_error: str):
        if token_count <= 0 or parser.is_at_end_or_line_end_offset(0):
            raise BarkError(line, empty_error)
        first = parser.peek_at(0)
        if first.is_(TokenType.IDENTIFIER) and token_count == 1:
            word = parser.normalise(first.value)
            if word in Keywords.BREAK_KEYWORDS:
                parser.advance_by(1)
                return [AstNode.Break(line)]
            if word in Keywords.CONTINUE_KEYWORDS:
                parser.advance_by(1)
                return [AstNode.Continue(line)]
        if parser.line_has_word(Keywords.PRINT_KEYWORDS):
            return [parser.get_print_parser().parse_print_for_next_tokens(token_count)]
        raise BarkError(line, empty_error)
